### PR TITLE
Opacity for legacy subTool GUI

### PR DESCRIPTION
**Pull request by krzychu124**
- UI opacity slider does not work for `legacySubTools`,
- eliminated possible memory leaks

### CODE REVIEW

**Review by kvakvs — APPROVED** [review]
Nice!

**Comment by chameleon-tbn** [comment]
it is still inconsistent imho. the speed limit signs in the UI should be same transparent as the main TMPE UI buttons.







additional the lane connectors are not getting transparent.



same for junction restrictions



And... sliders should be at 50% in the middle....

**Comment by kianzarrin** [comment]
can you give an example of legacy subtool that is affected by this?

**Comment by kianzarrin** [comment]
is this back-end only?

**Comment by krzychu124** [comment]
> can you give an example of legacy subtool that is affected by this?

Basically any tool with OnGUI UI (does not have LegacySubTool base), e.g. traffic lights tool, vehicle restrictions window, etc.

File: TLM/TLM/UI/LegacySubTool.cs [hunk 1]
```diff
@@ -183,5 +203,20 @@ public abstract class LegacySubTool {
                 GUI.DragWindow();
             }
         }
+
+        public void OnUpdate(ModUI.UIOpacityNotification subject) {
+            Texture2D windowTexture = windowTexture_;
```
Comment (DaEgi01) [inline TLM/TLM/UI/LegacySubTool.cs:208]: Can you check if just
UnityEngine.Object.Destroy(windowTexture_);
windowTexture_ = null;
would do too?

File: TLM/TLM/UI/LegacySubTool.cs [hunk 2]
```diff
@@ -152,6 +157,21 @@ public abstract class LegacySubTool {
...
         }
 
+        public virtual void OnDestroy() {
+            if (borderlessTexture_) {
+                UnityEngine.Object.Destroy(borderlessTexture_);
+                borderlessTexture_ = null;
+            }
+            borderlessStyle_ = null;
+
+            if (windowTexture_) {
```
Comment (DaEgi01) [inline TLM/TLM/UI/LegacySubTool.cs:167]: Is the check necessary?

File: TLM/TLM/UI/LegacySubTool.cs [hunk 3]
```diff
@@ -183,5 +203,20 @@ public abstract class LegacySubTool {
...
+            Texture2D windowTexture = windowTexture_;
+            windowTexture_ = null;
+            windowStyle_ = null; // rebuild style with new window texture
+            if (windowTexture) {
+                UnityEngine.Object.Destroy(windowTexture);
+            }
+            Texture2D borderlessTexture = borderlessTexture_;
+            borderlessTexture_ = null;
+            borderlessStyle_ = null; // rebuild style with new borderless texture
+            if (borderlessTexture) {
```
Comment (DaEgi01) [inline TLM/TLM/UI/LegacySubTool.cs:217]: Check necessary?

**Review by DaEgi01 — CHANGES_REQUESTED** [review]

**Comment by kianzarrin** [comment]
I tested this and confirm that it works.

**Review by kianzarrin — APPROVED** [review]

File: TLM/TLM/UI/LegacySubTool.cs (hunk 1 above)
Comment (krzychu124) [reply TLM/TLM/UI/LegacySubTool.cs:208]: Game engine will throw error if you try destroying already destroyed object, that's why is that ugly

**Comment by krzychu124** [comment]
@chameleon-tbn I addressed all requested changes (skipped speed limits tool, because it will be replaced anyways).
If you are edit lane connections, transparency will be ignored for active node, same for junction restrictions otherwise it should take the setting into account 😉

**Comment by chameleon-tbn** [comment]
Over all it looks OK. Overlay option activated all is faded. But for non-active junction if lane connector is selected, the transparency is not working

image

**Comment by chameleon-tbn** [comment]
Faster than Flash our @krzychu124

**Review by chameleon-tbn — APPROVED** [review]
Works as expected now

**Pull request merged**